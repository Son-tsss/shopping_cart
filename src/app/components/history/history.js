import React from 'react'
import PropTypes from 'prop-types'
import {connect} from 'react-redux'
import {bindActionCreators} from 'redux'

import {shoppingCartWithHistoryActions} from '../../actions'

const{undo, redo} = shoppingCartWithHistoryActions

const mapStateToProps = (state) => state

const mapDispatchToProps = (dispatch) => {
    return bindActionCreators({undo,redo}, dispatch)
}

const History = ({undo,redo}) => {
    return <div className="history">
        <button className="history-undo" onClick={undo}>UNDO</button>
        <button className="history-redo" onClick={redo}>REDO</button>
    </div>
}

History.propTypes = {
    undo: PropTypes.func.isRequired,
    redo: PropTypes.func.isRequired
}

export default connect(mapStateToProps, mapDispatchToProps)(History)