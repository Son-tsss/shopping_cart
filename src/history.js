const history = (reducer) => {
    const initialState = {
        previousStates: [],
        currentState: reducer(undefined, {}),
        futureStates: []
    }

    return (state = initialState, action) => {
        const {previousStates, currentState, futureStates} = state
        switch (action.type) {
            case 'UNDO':
                return {
                    previousStates: previousStates.slice(0, previousStates.length - 1),
                    currentState: previousStates[previousStates.length - 1],
                    futureStates: [...futureStates, currentState]
                }
            case 'REDO':
                return {
                    previousStates: [...previousStates, currentState],
                    currentState: futureStates[futureStates.length - 1],
                    futureStates: futureStates.slice(0, futureStates.length - 1)
                }
            default:
                return {
                    previousStates: [...previousStates, currentState],
                    currentState: reducer(currentState, action),
                    futureStates: []
                }
        }
    }
}

export default history