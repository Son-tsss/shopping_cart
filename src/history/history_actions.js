import {UNDO, REDO} from './history_action_types'

const undo = () => {
    const type = UNDO
    return {type}
}


const redo = () => {
    const type = REDO

    return {type}
}

export const historyActions = (actions) => {
    return {
        ...actions,
        undo,
        redo
    }
}