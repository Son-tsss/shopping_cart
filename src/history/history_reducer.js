import {UNDO, REDO} from './history_action_types'

const initialState = {
    states: [{}],
    currentStatePointer: -1
}

const historyReducer = (reducer) => {

    return (state = initialState, action) => {
        const {states, currentStatePointer} = state
        switch (action.type) {
            case UNDO:
                return {
                    currentStatePointer: currentStatePointer > 0 ? currentStatePointer - 1 : 0
                }
            case REDO:
                return {
                    currentStatePointer: currentStatePointer < states.length - 1
                        ? currentStatePointer + 1
                        : states.length - 1
                }
            default:
                return {
                    states: [...states.slice(0, currentStatePointer+1), reducer(states[currentStatePointer], action)],
                    currentStatePointer: currentStatePointer + 1
                }
        }
    }
}

export default historyReducer