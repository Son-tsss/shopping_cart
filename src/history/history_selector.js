export default (selector) => {
    return ({states, currentStatePointer}) => {
        const currentState = states[currentStatePointer]
        return {
            ...selector(currentState),
            canUndo: currentStatePointer > 0,
            canRedo: (currentStatePointer + 1) < states.length
        }
    }
}