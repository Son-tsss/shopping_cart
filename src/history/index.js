import reducer from './history_reducer'
import selector from './history_selector'
import {historyActions as actions} from './history_actions'

export default {
    reducer,
    selector,
    actions
}