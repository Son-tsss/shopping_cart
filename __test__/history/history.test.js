import history from "../../src/history/history_reducer"

const reducer = (state = 0, action) => {
    switch (action.type) {
        case "INCR":
            return state + 1
        default:
            return state
    }
}

it("should return reducer", () => {
    expect(history(reducer)).toBeInstanceOf(Function)
})

it("should have initial state of given reducer as previous state and correct current state", () => {
    const reducerWithHistory = history(reducer)
    const real = reducerWithHistory(undefined, {type: "INCR"})

    expect(real).toMatchSnapshot()
})

it("should have new state of given reducer as current state", () => {
    const reducerWithHistory = history(reducer)

    const initialState = {
        previousStates: [0, 1],
        currentState: 2,
        futureStates: []
    }

    const real = reducerWithHistory(initialState,{type: "INCR"})

    expect(real).toMatchSnapshot()
})

it("should have prev state as current and current state in futureStates after undo", () => {
    const reducerWithHistory = history(reducer)

    const initialState = {
        previousStates: [0,1],
        currentState: 2,
        futureStates: []
    }
    const real = reducerWithHistory(initialState,{type: "UNDO"})

    expect(real).toMatchSnapshot()
})

it("should have next state as current and current state in previousStates after redo", () => {
    const reducerWithHistory = history(reducer)

    const initialState = {
        previousStates: [0,1],
        currentState: 2,
        futureStates: [4, 3]
    }
    let real = reducerWithHistory(initialState,{type: "REDO"})

    const expected = {
        previousStates: [0, 1, 2],
        currentState: 3,
        futureStates: [4]
    }

    expect(real).toMatchSnapshot()
})

it("should have next state empty after new changes", () => {
    const reducerWithHistory = history(reducer)

    const initialState = {
        previousStates: [0,1],
        currentState: 2,
        futureStates: [4, 3]
    }

    let real = reducerWithHistory(initialState, {type: "INCR"})

    expect(real).toMatchSnapshot()
})