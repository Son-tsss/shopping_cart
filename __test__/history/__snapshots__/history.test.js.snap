// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should have initial state of given reducer as previous state and correct current state 1`] = `
Object {
  "currentState": 1,
  "futureStates": Array [],
  "previousStates": Array [
    0,
  ],
}
`;

exports[`should have new state of given reducer as current state 1`] = `
Object {
  "currentState": 3,
  "futureStates": Array [],
  "previousStates": Array [
    0,
    1,
    2,
  ],
}
`;

exports[`should have next state as current and current state in previousStates after redo 1`] = `
Object {
  "currentState": 3,
  "futureStates": Array [
    4,
  ],
  "previousStates": Array [
    0,
    1,
    2,
  ],
}
`;

exports[`should have next state empty after new changes 1`] = `
Object {
  "currentState": 3,
  "futureStates": Array [],
  "previousStates": Array [
    0,
    1,
    2,
  ],
}
`;

exports[`should have prev state as current and current state in futureStates after undo 1`] = `
Object {
  "currentState": 1,
  "futureStates": Array [
    2,
  ],
  "previousStates": Array [
    0,
  ],
}
`;
